### ISSUE TITLE
VS crash with IronPython and Dev 2015-08-21

### ISSUE THREAD

**Issue by andygraham** [Microsoft/PTVS#802]
VS 2013 Desktop Express, IronPython 2.7.5, PTVS Dev 2015-08-21 VS 2013

I spoke too soon :( I just got a hang while rebuilding my solution which is a mix of one IronPython project and three Windows Forms C# projects.  

PTVS Dev 2015-08-21 VS 2013 is definitely more stable than PTVS 2.2. I haven't yet encountered a crash or "Exception caused by an extension" message while editing the IronPython code which used to happen quite often in 2.2. I often got the crash below while rebuilding my solution in 2.2 and this is the first since installing Dev 2015-08-21 so somewhere, hiding in a dark murky corner there must be one or more unchecked casts that got overlooked when the rest were fixed!

Microsoft Visual Studio Express 2013 for Windows Desktop has stopped working.
Problem signature:
  Problem Event Name:	CLR20r3
  Problem Signature 01:	WDExpress.exe
  Problem Signature 02:	12.0.40629.0
  Problem Signature 03:	5590c29f
  Problem Signature 04:	Microsoft.PythonTools.IronPython.Interpreter
  Problem Signature 05:	2.2.30821.1
  Problem Signature 06:	55d7b607
  Problem Signature 07:	21a
  Problem Signature 08:	0
  Problem Signature 09:	System.InvalidCastException
  OS Version:	6.1.7601.2.1.0.768.3
  Locale ID:	2057
  Additional Information 1:	0a9e
  Additional Information 2:	0a9e372d3b4ad19135b953a78882e789
  Additional Information 3:	0a9e
  Additional Information 4:	0a9e372d3b4ad19135b953a78882e789

**Comment by zooba** [comment]
Hopefully there's a second event in your Event Log near that one with a stack trace. (Generally true, but especially for our IronPython support, stack traces are *very* helpful.)

**Comment by andygraham** [comment]
I found ActivityLog.xml in AppData\Roaming\Microsoft\WDExpress\12.0 with the following stack trace at the end with the source given as "Editor or Editor Extension" but the date looks wrong. It is 2015/08/22 and so is probably from PTVS 2.2 as I installed Dev 2015-08-21 on 2015/08/25 and the crash was on 2015/09/01

System.InvalidCastException: Unable to cast object of type 'IronPython.Runtime.Types.BuiltinFunction' to type 'IronPython.Runtime.Types.PythonType'. at Microsoft.IronPythonTools.Interpreter.RemoteInterpreter.<>c__DisplayClass64_0.<GetPythonTypeMemberType>b__0() at Microsoft.IronPythonTools.Interpreter.RemoteInterpreter.CallAndHandle[T](Func`1 action, T defaultValue) at Microsoft.IronPythonTools.Interpreter.RemoteInterpreter.GetPythonTypeMemberType(ObjectIdentityHandle pythonType) at Microsoft.IronPythonTools.Interpreter.RemoteInterpreterProxy.GetPythonTypeMemberType(ObjectIdentityHandle value) at Microsoft.IronPythonTools.Interpreter.RemoteInterpreterProxy.GetPythonTypeMemberType(ObjectIdentityHandle value) at Microsoft.IronPythonTools.Interpreter.IronPythonType.get_MemberType() at Microsoft.PythonTools.Analysis.Values.BuiltinClassInfo.get_MemberType() at Microsoft.PythonTools.Analysis.Values.SpecializedNamespace.get_MemberType() at Microsoft.PythonTools.Analysis.MemberResult.GetMemberType() at Microsoft.PythonTools.Analysis.MemberResult.get_MemberType() at Microsoft.PythonTools.Intellisense.CompletionAnalysis.PythonCompletion(IGlyphService service, MemberResult memberResult) at Microsoft.PythonTools.Intellisense.NormalCompletionAnalysis.<>c__DisplayClass5_0.<GetCompletions>b__3(MemberResult m) at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext() at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext() at System.Linq.Buffer`1..ctor(IEnumerable`1 source) at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__0.MoveNext() at Microsoft.VisualStudio.Language.Intellisense.BulkObservableCollection`1.AddRange(IEnumerable`1 items) at Microsoft.PythonTools.Intellisense.FuzzyCompletionSet..ctor(String moniker, String displayName, ITrackingSpan applicableTo, IEnumerable`1 completions, CompletionOptions options, IComparer`1 comparer, Boolean matchInsertionText) at Microsoft.PythonTools.Intellisense.NormalCompletionAnalysis.GetCompletions(IGlyphService glyphService) at Microsoft.PythonTools.Intellisense.CompletionSource.AugmentCompletionSession(ICompletionSession session, IList`1 completionSets) at Microsoft.VisualStudio.Language.Intellisense.Implementation.CompletionSession.Start() at Microsoft.VisualStudio.Language.Intellisense.Implementation.CompletionBroker.TriggerCompletion(ITextView textView) at Microsoft.PythonTools.Intellisense.IntellisenseController.TriggerCompletionSession(Boolean completeWord, Nullable`1 commitByDefault) at Microsoft.PythonTools.Intellisense.IntellisenseController.HandleChar(Char ch) at Microsoft.PythonTools.Intellisense.IntellisenseController.Exec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut) at Microsoft.VisualStudio.Editor.Implementation.CommandChainNode.InnerExec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut) at Microsoft.VisualStudio.Editor.Implementation.SimpleTextViewWindow.Exec(Guid& pguidCmdGroup, UInt32 nCmdID, UInt32 nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut) at Microsoft.VisualStudio.Editor.Implementation.VsKeyboardFilter.SendCommand(Guid cmdGroup, UInt32 cmdID, Object inParam) at Microsoft.VisualStudio.Editor.Implementation.VsKeyboardFilter.TextInput(TextCompositionEventArgs args) at Microsoft.VisualStudio.Text.Editor.Implementation.KeyProcessorDispatcher.<DispatchTextInputEvents>b__0(KeyProcessor p, TextCompositionEventArgs args) at Microsoft.VisualStudio.Text.Editor.Implementation.KeyProcessorDispatcher.<>c__DisplayClass1e`1.<Dispatch>b__1a() at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call)

**Comment by andygraham** [comment]
I found this in the computer system application logs. The date for this one is correct so this should be the crash I observed. Owing to the xml it doesn't preview properly when pasted. I think I corrected it for the first few lines which are the important one, later ones may be corrupted.

Application: WDExpress.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.InvalidCastException
Stack:
   at Microsoft.IronPythonTools.Interpreter.RemoteInterpreter.<>c__DisplayClass74_0.\<GetBuiltinFunctionDocumentation>b__0()
   at Microsoft.IronPythonTools.Interpreter.RemoteInterpreter.CallAndHandle(Func`1 action)
   at Microsoft.IronPythonTools.Interpreter.RemoteInterpreter.GetBuiltinFunctionDocumentation(ObjectIdentityHandle function)
   at Microsoft.IronPythonTools.Interpreter.RemoteInterpreterProxy.GetBuiltinFunctionDocumentation(ObjectIdentityHandle value)
   at Microsoft.IronPythonTools.Interpreter.RemoteInterpreterProxy.GetBuiltinFunctionDocumentation(Microsoft.IronPythonTools.Interpreter.ObjectIdentityHandle)
   at Microsoft.IronPythonTools.Interpreter.IronPythonBuiltinFunction.get_Documentation()
   at Microsoft.PythonTools.Analysis.Values.BuiltinFunctionInfo.get_Documentation()
   at Microsoft.PythonTools.Analysis.Values.SpecializedNamespace.get_Documentation()
   at Microsoft.PythonTools.Analysis.MemberResult.get_Documentation()
   at Microsoft.PythonTools.Intellisense.CompletionAnalysis+<>c__DisplayClass13_0.<PythonCompletion>b__0()
   at Microsoft.PythonTools.Intellisense.DynamicallyVisibleCompletion.get_Description()
   at Microsoft.VisualStudio.Language.Intellisense.Implementation.DefaultCompletionSetPresenter.ShowToolTip()
   at Microsoft.VisualStudio.Language.Intellisense.Implementation.DefaultCompletionSetPresenter.OnShowTooltipTimer_Tick(System.Object, System.EventArgs)
   at System.Windows.Threading.DispatcherTimer.FireTick(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)

**Comment by zooba** [comment]
Looks fine to me. I normally find/replace ` at ` with `\n at ` when copying out of the activity log. Putting it in triple backquotes (I just updated your comment with them) also makes it easier to read by disabling wrapping.

And we didn't actually go through and check all the casts - we did some manual static analysis and discovered a point where we were putting the wrong type in and that was causing (so we thought) all the issues.

This one is strange, since there would have been an earlier call that immediately triggered the same exception (getting the name of the function). I think what we actually have here is a race condition caused by the rebuild.

I can't finish fixing everything right now, so note to myself (or whoever picks this up): `PythonObject` probably should capture the remote interpreter so that it doesn't call through to the wrong instance after it gets recreated.
